frontend/dashboard/components/NewApplicationForm/NewApplicationForm.tsx

```diff
@@ -0,0 +1,144 @@
+import React, { type FormEvent, type ChangeEvent } from 'react';
+import classes from './NewApplicationForm.module.css';
+import { StudioButton, StudioSpinner } from '@studio/components';
+import { useTranslation } from 'react-i18next';
+import { ServiceOwnerSelector } from '../ServiceOwnerSelector';
+import { RepoNameInput } from '../RepoNameInput';
+import { type User } from 'app-shared/types/Repository';
+import { type Organization } from 'app-shared/types/Organization';
+import { useSelectedContext } from '../../hooks/useSelectedContext';
+import { SelectedContextType } from 'app-shared/navigation/main-header/Header';
+import { type NewAppForm } from '../../types/NewAppForm';
+import { useCreateAppFormValidation } from './hooks/useCreateAppFormValidation';
+import { Link } from 'react-router-dom';
+
+type CancelButton = {
+  onClick: () => void;
+  type: 'button';
+};
+
+type CancelLink = {
+  href: string;
+  type: 'link';
+};
+export type CancelComponent = CancelButton | CancelLink;
+
+export type NewApplicationFormProps = {
+  onSubmit: (newAppForm: NewAppForm) => Promise<void>;
+  user: User;
+  organizations: Organization[];
+  isLoading: boolean;
+  submitButtonText: string;
+  formError: NewAppForm;
+  setFormError: React.Dispatch<React.SetStateAction<NewAppForm>>;
+  cancelComponent: CancelComponent;
+};
+
+export const NewApplicationForm = ({
+  onSubmit,
+  user,
+  organizations,
+  isLoading,
+  submitButtonText,
+  formError,
+  setFormError,
+  cancelComponent,
+}: NewApplicationFormProps): React.JSX.Element => {
+  const { t } = useTranslation();
+  const selectedContext = useSelectedContext();
+  const { validateRepoOwnerName, validateRepoName } = useCreateAppFormValidation();
+
+  const defaultSelectedOrgOrUser: string =
+    selectedContext === SelectedContextType.Self || selectedContext === SelectedContextType.All
+      ? user.login
+      : selectedContext;
+
+  const validateTextValue = (event: ChangeEvent<HTMLInputElement>) => {
+    const { errorMessage: repoNameErrorMessage, isValid: isRepoNameValid } = validateRepoName(
+      event.target.value,
+    );
+    setFormError((previous) => ({
+      ...previous,
+      repoName: isRepoNameValid ? '' : repoNameErrorMessage,
+    }));
+  };
+
+  const handleSubmit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
+    event.preventDefault();
+
+    const formData: FormData = new FormData(event.currentTarget);
+    const newAppForm: NewAppForm = {
+      org: formData.get('org') as string,
+      repoName: formData.get('repoName') as string,
+    };
```

Comment:
We can simplify this code by using `Object.entries`, what do you think? 😊 
```suggestion
    const newAppForm = Object.entries(formData) as NewAppForm ;
```

Reply:
Smart! 🤩

Reply:
For some reason, this change makes 10 unit tests fail. It says that the button was not clicked 🤔

Reply:
Ah, makes sense since my example was not 100% correct, it should be `Object.fromEntries` not `Object.entries`. 😊

Reply:
hmmm 🤔 
Still fails, and now on 13 tests 😱

Reply:
Give me a call tomorrow, will you? 😊